### ISSUE TITLE
Script exploits

### ISSUE THREAD

**Issue by mateon1** [danielyxie/bitburner#164]
## Avoid ram usage for function calls:
```javascript
hack_func = hack;
hack_func("hostname");
```

## Avoid ram usage for while:
```javascript
while (cond) {
  stuff;
}
```
turns into
```javascript
function recursive() {
  stuff;
  cond && recursive();
}
```

## Execute arbitrary Javascript [with no RAM cost]
```javascript
hack.constructor("alert(1)")();
```

---

Also, there are some annoying limitations.
It's impossible to use anything like `[1, 2, 3].map(function a(x) { return x * 2; })` [function expressions not implemented], or `function mapper(x) { return x * 2; }; [1,2,3].map(mapper);` [function declarations are not actual Javascript functions]

**Comment by danielyxie** [comment]
Thanks, this is really helpful as I'm still not that experienced with JavaScript

**Comment by danielyxie** [comment]
"Avoid RAM usage for function" and "execute arbitrary javascript" exploits fixed in v0.34.2

**Comment by danielyxie** [comment]
Fixed with new RAM calculation mechanic. Thanks!
